### PR TITLE
Translate `_d_arrayappend{T,cTX}` to templates

### PR DESCRIPTION
**Pull request by teodutu**
This PR lowers expressions such as `a ~= b` to:
```d
_d_arrayappendcTX(a, 1), a[a.length - 1] = b, a;
// Note that the assignment above is actually a construction.
```
when `b` is a single element, or
```d
_d_arrayappendT(a, b);
```
when `b` is an array. These lowerings are now moved to the compiler's frontend, as opposed to e2ir.d.

This revives https://github.com/dlang/dmd/pull/9982.

### CODE REVIEW

File: src/dmd/e2ir.d [hunk 1]
```diff
@@ -2964,6 +2908,19 @@ extern (C++) class ToElemVisitor : Visitor
...
+            {
+                elem *e = toElem(ce.e2, irs);
+                if (irs.params.cov && ce.e2.loc.linnum)
+                    e = el_combine(incUsageElem(irs, ce.e2.loc), e);
+                if (tybasic(e.Ety) == TYstruct)
+                    e.ET = Type_toCtype(ce.e2.type);
+                elem_setLoc(e, ce.loc);
+                result = e;
+                return;
+            }
```
Comment (ibuclaw) [inline src/dmd/e2ir.d:2922]: This doesn't look relevant to the change, and the backend should already optimize this case.

File: src/dmd/expression.d [hunk 2]
```diff
@@ -6161,6 +6162,7 @@ extern (C++) final class UshrAssignExp : BinAssignExp
  */
 extern (C++) class CatAssignExp : BinAssignExp
 {
+    bool isCTFEHookPart;
```
Comment (ibuclaw) [inline src/dmd/expression.d:6165]: This has not been updated in the headers.

File: src/dmd/expression.d [hunk 3]
```diff
@@ -6612,7 +6614,8 @@ extern (C++) final class CondExp : BinExp
 
                     if (v.needsScopeDtor())
                     {
-                        if (!vcond)
+                        auto cve = ce.econd.isVarExp();
+                        if (!vcond && !cve)
```
Comment (ibuclaw) [inline src/dmd/expression.d:6618]: Doesn't seem relevant here either.

File: src/dmd/s2ir.d [hunk 4]
```diff
@@ -184,6 +184,31 @@ private extern (C++) class S2irVisitor : Visitor
...
+         *          b;
+         * Since __ctfe is always false at runtime, the if body can be omitted,
+         * and the statement rewritten simply as `b;`.
+         *
+         * This is necessary for expressions such as `a ~= b`, where `b` is not
+         * an array. In expressionsem.d, they are lowered to:
+         * `__ctfe ? a ~= b : _d_arrayappendcTX(a, 1), a[$ - 1] = b, a;`
+         * If CTFE hasn't already interpreted `a ~= b`, then this branch must be
+         * removed, as e2ir.d no longer supports the runtime hook to
+         * `_d_arrayappendcTX`.
```
Comment (ibuclaw) [inline src/dmd/s2ir.d:201]: Can't you just remove it during semantic?

**Review by ibuclaw — DISMISSED** [review]

File: src/dmd/e2ir.d (hunk 1 above)
Comment (teodutu) [reply src/dmd/e2ir.d:2922]: If I don't do this here, the `true` branch of the cond reaches `visit(CatExp)`, which is triggers these `assert`s [1] [2]. In addition, it doesn't make sense to lowert the `CatExp` twice: in expressionsem.d and here.

[1] https://github.com/dlang/dmd/blob/3f1b2c4e2d7b3a717e8e0158cd4b6731b04e28ad/src/dmd/e2ir.d#L2717
[2] https://github.com/dlang/dmd/blob/3f1b2c4e2d7b3a717e8e0158cd4b6731b04e28ad/src/dmd/e2ir.d#L2722

File: src/dmd/s2ir.d (hunk 4 above)
Comment (teodutu) [reply src/dmd/s2ir.d:201]: I'm not sure I understand what you mean. I create the lowering during semantic3 and the lowering is needed for CTFE. I don't see where in semantic i should remove it.

File: src/dmd/expression.d (hunk 3 above)
Comment (teodutu) [reply src/dmd/expression.d:6618]: This was needed because it caused errors when copying `__ctfe` to a temporary variable. I could have simply excluded that case, but I chose the more general approach, where I excluded every simple variable, as they don't need to be saved as temporaries.

Another approach would have been to use `extractSideEffect` instead of `copyToTemp` in the code below. `extractSideEffect` would leave the cond unchanged if it's trivial and has no side effects. Would this work better?

File: src/dmd/expression.d (hunk 2 above)
Comment (teodutu) [reply src/dmd/expression.d:6165]: I've added it to `expression.h`.

**Comment by radcapricorn** [comment]
> This PR lowers expressions such as `a ~= b` to:
> 
> ```d
> __ctfe ? _d_arrayappendcTX(a, 1), a[a.length - 1] = b, a;
> ```
> 
> when `b` is a single element

I don't think that is a correct lowering. `a ~= b` should be appending a copy of `b` to `a` when `b` is lvalue or emplacing `b` at the end of `a` when `b` is rvalue. As worded, however, this looks like assignment of `b` to a (default-initialized?) newly added element of `a`, which may not even be legal (i.e. disabled default constructor). What's the actual semantics here?

**Comment by teodutu** [comment]
> > This PR lowers expressions such as `a ~= b` to:
> > ```d
> > __ctfe ? _d_arrayappendcTX(a, 1), a[a.length - 1] = b, a;
> > ```  
> > when `b` is a single element
> 
> I don't think that is a correct lowering. `a ~= b` should be appending a copy of `b` to `a` when `b` is lvalue or emplacing `b` at the end of `a` when `b` is rvalue. As worded, however, this looks like assignment of `b` to a (default-initialized?) newly added element of `a`, which may not even be legal (i.e. disabled default constructor). What's the actual semantics here?

First, thanks for pointing out to me that I had forgotten the `true` branch of the cond. The correct lowering should be:
```d
__ctfe ? a ~= b : _d_arrayappendcTX(a, 1), a[a.length - 1] = b, a;
```
I've also updated the PR's description.

Now to answer your question, the lowering creates a `ConstructExp` for the assignment [1], which, differentiates between lvalue and rvalue rhs's in e2ir.d.

[1] https://github.com/dlang/dmd/blob/92d463064b567dd2e0a88aba2d32117a65be47d6/src/dmd/expressionsem.d#L10292

**Comment by radcapricorn** [comment]
Could you then reword the description please? Because `_d_arrayappendcTX(a, 1), a[a.length - 1] = b` still reads "grow array, assign at end", not "construct at end".

**Comment by teodutu** [comment]
Done @radcapricorn. Is the description clearer now? I'd prefer to keep the code itself as it is, because that's more or less what you get if you print `result.toChars()`, after running semantic on the lowering.

File: src/dmd/s2ir.d (hunk 4 above)
Comment (ibuclaw) [reply src/dmd/s2ir.d:201]: Sounds like something has gone wrong to me.  Lowering should only occur _after_ CTFE has had a chance to reduce the code, not before.

File: src/dmd/dinterpret.d [hunk 5]
```diff
@@ -4878,6 +4878,10 @@ public:
 
                 return;
             }
+            else if (fd.ident == Id._d_arrayappendT)
+                assert(0, "CTFE cannot interpret _d_arrayappendT!");
+            else if (fd.ident == Id._d_arrayappendcTX)
+                assert(0, "CTFE cannot interpret _d_arrayappendcTX!");
```
Comment (ibuclaw) [inline src/dmd/dinterpret.d:4884]: Having another look over, if I understand this function right, you should instead implement these two functions before CTFE _can_ interpret arrayappend operations.  Then there's no need for the `(__ctfe) ? a : b` set-up you've done in semantic, just generate the lowering to template directly.

Comment (teodutu) [reply src/dmd/dinterpret.d:4884]: If the lowering reaches CTFE, this error [1] is triggered because of the mixin in the hook's body [2]. That's why those `assert`s are here and that's why I use `__ctfe ? ...`.

[1] https://github.com/dlang/dmd/blob/b14861d42ce3da9cf789988621c46dba1102ee0c/src/dmd/dinterpret.d#L4967-L4972
[2] https://github.com/dlang/druntime/blob/cc101972b1f6d3047ca7b54cbf4aee11ac499b75/src/core/internal/array/appending.d#L42

File: src/dmd/s2ir.d (hunk 4 above)
Comment (RazvanN7) [reply src/dmd/s2ir.d:201]: The problem is that when the compiler analyzes the body of a function it cannot know if the function is going to be interpreted or not in the future. Since we cannot interpret `_d_arrayappend` this trick with `__ctfe` is used.

Comment (teodutu) [reply src/dmd/s2ir.d:201]: Performing the lowering in e2ir.d, where it previously was, has the following disadvantages:
1. Maintainability decreases as LDC and GDC will have to implement the lowering themselves.
2. Readability also decreases, because the code in e2ir.d is less expressive than that in expressionsem.d.

While it is true that lowering the `CatAssign` as early as expressionsem.d makes the logic more complex, requiring that e2ir.d and s2ir.d omit the `true` branch of the `__ctfe` cond, I believe this approached is preferable because it elegantly solves the problems above, while also clearly separating the expressions intended for CTFE and for code generation.

Comment (ibuclaw) [reply src/dmd/s2ir.d:201]: > I believe this approach is preferable because it elegantly solves the problems above, while also clearly separating the expression intended for CTFE from those meant for code generation.

No it doesn't, you're adding semantic logic to the code generator.  The code generator should at best be just a trivial translation from one representation to another.

What I see is that you've added some new `assert(0)` branches in the code generator. This means that the front-end should _never_ leak these AST constructs to the back-end, and yet, that is exactly what the lowering does, so the code generator is becomes even more complex than it needs to be by having a need to filter AST nodes that should have not have reached it in the first place.

Comment (RazvanN7) [reply src/dmd/s2ir.d:201]: Hmm, maybe we can make the hook interpret-able. That would get rid of the filtering code in `e2ir.d` and would still enable us to perform the lowering in the frontend. Is that an acceptable solution @ibuclaw ?

**Comment by teodutu** [comment]
In order to pass the tests, this PR requires an updated implementation of `_d_arrayappendT`: https://github.com/dlang/druntime/pull/3805.

File: src/dmd/s2ir.d [hunk 6]
```diff
@@ -184,6 +184,7 @@ private extern (C++) class S2irVisitor : Visitor
         block *bexit = mystate.breakBlock ? mystate.breakBlock : dmd.backend.global.block_calloc();
 
         incUsage(irs, s.loc);
+
```
Comment (RazvanN7) [inline src/dmd/s2ir.d:187]: unrelated change, please undo.

File: test/compilable/extra-files/vcg-ast.d.cg [hunk 7]
```diff
@@ -140,10 +140,104 @@ values!(__c_wchar_t)
 	pure nothrow @safe __c_wchar_t[] values()
 	{
 		__c_wchar_t[] values = null;
-		values ~= cast(__c_wchar_t)'\uffff';
```
Comment (RazvanN7) [inline test/compilable/extra-files/vcg-ast.d.cg:143]: Just get rid of `~=` in the initial test. The [PR that introduced](https://github.com/dlang/dmd/pull/10155) this fixed a bug report that can be reproduced without the cat assign operator. What is problematic is the `cast(__c_wchar_t)` that should still exist.

This will make it so you don't need to put generated code for _d_arrayappendcTXImpl in the test.

**Review by RazvanN7 — CHANGES_REQUESTED** [review]

File: src/dmd/expression.d [hunk 8]
```diff
@@ -6256,6 +6257,7 @@ extern (C++) final class UshrAssignExp : BinAssignExp
  */
 extern (C++) class CatAssignExp : BinAssignExp
 {
+    bool isCTFEHookPart;
```
Comment (dkorpel) [inline src/dmd/expression.d:6260]: The new field needs documentation

File: src/dmd/dinterpret.d [hunk 9]
```diff
@@ -6359,6 +6412,26 @@ public:
     {
         assert(0); // This should never be interpreted
     }
+
+    private CallExp isRuntimeHook(Expression e, Identifier id)
```
Comment (dkorpel) [inline src/dmd/dinterpret.d:6416]: New function needs documentation

File: src/dmd/dinterpret.d [hunk 10]
```diff
@@ -1038,6 +1038,13 @@ public:
         if (exceptionOrCant(e))
             return;
 
+        if (isRuntimeHook(s.exp, Id._d_arrayappendT) || isRuntimeHook(s.exp, Id._d_arrayappendTTrace))
```
[the remark this answers is not in the document]
Comment (dkorpel) [reply src/dmd/dinterpret.d:1041]: It should also have test coverage

File: src/dmd/expressionsem.d [hunk 11]
```diff
@@ -10177,6 +10177,129 @@ private extern (C++) final class ExpressionSemanticVisitor : Visitor
...
+
+            if (exp.op == EXP.concatenateAssign)
+            {
+                Identifier hook = global.params.tracegc ? Id._d_arrayappendTTrace : Id._d_arrayappendT;
+
+                if (!verifyHookExist(exp.loc, *sc, hook, "appending array to arrays", Id.object))
+                    return setError();
+
+                // Lower to object._d_arrayappendT{,Trace}({file, line, funcname}, e1, e2)
+                 Expression id = new IdentifierExp(exp.loc, Id.empty);
```
Comment (dkorpel) [inline src/dmd/expressionsem.d:10210]: Indentation needs one space less

**Review by dkorpel — COMMENTED** [review]
I've put a few nits, but my bigger issue is with the size. The added code is substantially more complex than the removed code. Especially seeing how the hook functions get rewritten back  to `CatAssignExp` for CTFE looks clumsy. I feel we can do better than this, but unfortunately I can't give an insightful direction without delving into the project myself, so at this moment I'm abstaining.

I'm curious about the following things:
- How many more hooks are there to be translated to templates?
- Do those require PRs of similar complexity?
- Are there 'clean up' PRs planned after this?

Maybe this was all planned out somewhere, but I can't find the document. (I'm looking at https://github.com/dlang/projects/issues/25)

**Comment by RazvanN7** [comment]
@dkorpel There really isn't anything too complex happening here. The extra lines that you are seeing are added because:

1. We need to redo the expression for interpretation in case of ctfe. This is done for all other hooks.
2. Constructing a callExp in the frontend is more verbose than doing it in the gluelayer because the function/field names are much more descriptive and therefore more verbose.

In addition, some code from previous commits is obsolete and therefore will be deleted so the size will get smaller.

The idea here is that we might add a bit of complexity but, in general, it will worth it.

**Comment by dkorpel** [comment]
> There really isn't anything too complex happening here. The extra lines that you are seeing are added because:
> 
>  1. We need to redo the expression for interpretation in case of ctfe. This is done for all other hooks.

Can't you interpret the hook? There's an explicit check, but it asserts:
```D
else if (fd.ident == Id._d_arrayappendcTX)
    assert(0, "CTFE cannot interpret _d_arrayappendcTX!");
```

**Comment by dkorpel** [comment]
> What do you mean? Anyone is welcome to improve the code is possible.

I wondered if some of this code is just scaffolding that can be removed once the translation is done.

File: src/dmd/expressionsem.d [hunk 12]
```diff
@@ -10177,6 +10177,129 @@ private extern (C++) final class ExpressionSemanticVisitor : Visitor
...
+                arguments.push(new IntegerExp(exp.loc, 1, Type.tsize_t));
+
+                Expression ce = new CallExp(exp.loc, id, arguments);
+                ce = ce.expressionSemantic(sc);
+
+                Expression eValue2;
+                Expression value2 = exp.e2;
+                if (!value2.isVarExp && !value2.isIntegerExp &&
+                    !value2.isStructLiteralExp && !value2.isArrayLiteralExp &&
+                    !value2.isAssocArrayLiteralExp)
```
Comment (dkorpel) [inline src/dmd/expressionsem.d:10281]: What is this checking for?

**Comment by dkorpel** [comment]
> No, the hook ends up calling C code.

I mean instead of "AST node -> hook -> AST node -> special AST node interpret", do "AST node -> hook -> special hook interpret"

**Comment by RazvanN7** [comment]
Essentially, you would need to copy paste the code of the cat assign interpretation. So recreating the expression seems like a simpler alternative.

**Comment by RazvanN7** [comment]
> I wondered if some of this code is just scaffolding that can be removed once the translation is done.

Not that I know of.

**Comment by teodutu** [comment]
@dkorpel There are around 100 runtime hooks in total:
```
druntime/src $ egrep -o -R "_d_.*\(.*[^;]$" | cut -d ':' -f 2 | egrep -o "_d_[^(]*" | sort -u
102  # excluding the already translated hooks and _d_arrayappend{T,cTX}
```

Out of these, some do not require translation (such as `_d_unittest`, for example), or their translation would bring no benefits (such as `_d_allocmemory`). After you factor out these hooks, you end up with ~25 hooks. However, many of them perform similar tasks and might be translated together (such as `_d_newarray{T,U,iT,mTX,miTX}`).

File: src/dmd/expression.d (hunk 8 above)
Comment (teodutu) [reply src/dmd/expression.d:6260]: I removed it.

File: src/dmd/expressionsem.d [hunk 13]
```diff
@@ -10177,6 +10177,129 @@ private extern (C++) final class ExpressionSemanticVisitor : Visitor
...
+                Identifier hook = global.params.tracegc ? Id._d_arrayappendTTrace : Id._d_arrayappendT;
+
+                if (!verifyHookExist(exp.loc, *sc, hook, "appending array to arrays", Id.object))
+                    return setError();
+
+                // Lower to object._d_arrayappendT{,Trace}({file, line, funcname}, e1, e2)
+                 Expression id = new IdentifierExp(exp.loc, Id.empty);
+                id = new DotIdExp(exp.loc, id, Id.object);
+                id = new DotIdExp(exp.loc, id, hook);
+                id = id.expressionSemantic(sc);
```
[the remark this answers is not in the document]
Comment (teodutu) [reply src/dmd/expressionsem.d:10213]: I removed this call and as many calls under `EXP.concatenateElemAssign` as I could. One call is still needed before creating the `CommaExp`. Otherwise, compilation raises a seg fault.

File: src/dmd/expressionsem.d (hunk 12 above)
Comment (teodutu) [reply src/dmd/expressionsem.d:10276]: When lowering an expression like `a ~= a[$ - 1]`, `$` is evaluated after calling `_d_arrayappendcTX`. The hook extends the array, thus changing its length and therefore leading to an incorrect evaluation of `$`.

This code forces every non-trivial expression to be saved to a temporary variable. Simply calling `extractSideEffect` would not work because it does not create temporaries for `IndexExp`s.

Comment (dkorpel) [reply src/dmd/expressionsem.d:10276]: The corresponding check in the original code:
```D
if (e2.Eoper != OPvar && e2.Eoper != OPconst)
                {
                    // Evaluate e2 and assign result to temporary s2.
                    // Do this because of:
                    //    a ~= a[$-1]
                    // because $ changes its value
```
Why is there no comment anymore, and why does it include struct/array literals? I think that goes wrong when doing something like `a ~= [ a[$-1] ];`.

Comment (teodutu) [reply src/dmd/expressionsem.d]: I found a better approach: I imitated the previous code from e2ir.d here.

File: src/dmd/expressionsem.d [hunk 14]
```diff
@@ -10183,6 +10183,130 @@ private extern (C++) final class ExpressionSemanticVisitor : Visitor
...
+                auto arguments = new Expressions();
+                arguments.reserve(5);
+                if (global.params.tracegc)
+                {
+                    auto funcname = (sc.callsc && sc.callsc.func) ? sc.callsc.func.toPrettyChars() : sc.func.toPrettyChars();
+                    arguments.push(new StringExp(exp.loc, exp.loc.filename.toDString()));
+                    arguments.push(new IntegerExp(exp.loc, exp.loc.linnum, Type.tint32));
+                    arguments.push(new StringExp(exp.loc, funcname.toDString()));
+                }
+                arguments.push(exp.e1);
```
Comment (RazvanN7) [inline src/dmd/expressionsem.d:10228]: newline before this statement please.

File: src/dmd/expressionsem.d [hunk 15]
```diff
@@ -10183,6 +10183,130 @@ private extern (C++) final class ExpressionSemanticVisitor : Visitor
...
+                arguments.push(exp.e2);
+
```
Comment (RazvanN7) [inline src/dmd/expressionsem.d:10230]: no newline needed.

File: src/dmd/expressionsem.d [hunk 16]
```diff
@@ -10183,6 +10183,130 @@ private extern (C++) final class ExpressionSemanticVisitor : Visitor
...
+                Expression ce = new CallExp(exp.loc, id, arguments);
+
+                *output = ce.expressionSemantic(sc);
```
Comment (RazvanN7) [inline src/dmd/expressionsem.d:10233]: no newline needed

File: src/dmd/expressionsem.d [hunk 17]
```diff
@@ -10183,6 +10183,130 @@ private extern (C++) final class ExpressionSemanticVisitor : Visitor
...
+            }
+            else if (exp.op == EXP.concatenateElemAssign)
+            {
+                if (auto ve = exp.e1.isVarExp)
```
Comment (RazvanN7) [inline src/dmd/expressionsem.d:10237]: A comment explaining this case would be useful.

File: src/dmd/expressionsem.d [hunk 18]
```diff
@@ -10183,6 +10183,130 @@ private extern (C++) final class ExpressionSemanticVisitor : Visitor
...
+                     * for expressions like `a ~= a[$-1]`. Here, $ will be modified
+                     * by calling `_d_arrayappendcT`, so we need to save `a[$-1]` in
+                     * a temporary variable.
+                     */
+                    value2 = extractSideEffect(sc, "__appendtmp", eValue2, value2, true);
+                    exp.e2 = value2;
+
+                    // `__appendtmp*` will be destroyed together with the array `exp.e1`.
+                    auto vd = eValue2.isDeclarationExp().declaration.isVarDeclaration();
+                    vd.storage_class |= STC.nodtor;
```
Comment (RazvanN7) [inline src/dmd/expressionsem.d:10293]: I don't think this is necessary. Isn't the length an unsigned integer type?

**Review by RazvanN7 — CHANGES_REQUESTED** [review]
Almost done. Mostly nitpicks.

**Comment by RazvanN7** [comment]
It seems you still haven't fixed the Azure failures.

File: src/dmd/expressionsem.d (hunk 18 above)
Comment (teodutu) [reply src/dmd/expressionsem.d:10293]: This is not the length. This is the entire rhs of the CatAssign expression.

**Review by RazvanN7 — APPROVED** [review]

**Comment by RazvanN7** [comment]
You need a rebase to latest master to get rid of the FreeBSD failure. Also, mind the other nits in my review.

**Comment by RazvanN7** [comment]
@dkorpel @ibuclaw please re-review.

**Review by dkorpel — COMMENTED** [review]
Are there no existing tests for global.params.tracegc?

File: src/dmd/dinterpret.d [hunk 19]
```diff
@@ -6370,6 +6431,31 @@ public:
     {
         assert(0); // This should never be interpreted
     }
+
+    /*********************************************
+     * Checks if the given expresion is a call to the runtime hook `id`.
+     * Params:
+     *    e = the expression to check
+     *    id = the identifier of the runtime hook
+     */
```
Comment (dkorpel) [inline src/dmd/dinterpret.d:6440]:
```suggestion
     * Returns: `e` cast to `CallExp` if it's the hook, `null` otherwise
     */
```

**Review by dkorpel — APPROVED** [review]

**Comment by teodutu** [comment]
@dkorpel There are tests for `global.params.tracegc` in DRuntime. One such test is in druntime/test/profile/src/profilegc.d.

**Comment by dkorpel** [comment]
> One such test is in druntime/test/profile/src/profilegc.d.

Good, so the codecov warnings are false positives.

**Comment by CyberShadow** [comment]
This pull request introduced a regression:
https://issues.dlang.org/show_bug.cgi?id=23965
